fiat/coingecko.go

```diff
@@ -0,0 +1,276 @@
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// CoingeckoDownloader stores CoinGecko API parameters
+type CoingeckoDownloader struct {
+	url                string
+	coin               string
+	periodSeconds      int
+	db                 *db.RocksDB
+	timeFormat         string
+	httpTimeoutSeconds time.Duration
+	test               bool
+}
+
+// NewCoingeckoDownloader initiallizes the downloader for CoinGecko API.
+// If the "test" flag is true, then downloader will sync only the last few days instead of the whole history.
+func NewCoingeckoDownloader(db *db.RocksDB, params string, test bool) (*CoingeckoDownloader, error) {
+	var cg = &CoingeckoDownloader{}
+	type coingeckoParams struct {
+		URL           string `json:"url"`
+		Coin          string `json:"coin"`
+		PeriodSeconds int    `json:"periodSeconds"`
+	}
+	cgParams := &coingeckoParams{}
+	err := json.Unmarshal([]byte(params), &cgParams)
+	if err != nil {
+		return nil, err
+	}
+	if cgParams.URL == "" || cgParams.PeriodSeconds == 0 {
+		return nil, errors.New("Missing parameters")
+	}
+	cg.timeFormat = "02-01-2006" // Layout string for CoinGecko date formatting (DD-MM-YYYY)
+	cg.httpTimeoutSeconds = 15 * time.Second
+	cg.url = cgParams.URL
+	cg.coin = cgParams.Coin
+	cg.periodSeconds = cgParams.PeriodSeconds // Time period for syncing the latest market data
+	cg.db = db
+	cg.test = test
+	return cg, err
+}
+
+// GetMarketData retrieves the response from coingecko API at the specified date.
+// If timestamp is nil, it fetches the latest market data available.
+func (cg *CoingeckoDownloader) GetMarketData(timestamp *time.Time) ([]byte, error) {
+	requestURL := cg.url + "/coins/" + cg.coin
+	if timestamp != nil {
+		requestURL += "/history"
+	}
+
+	req, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		glog.Errorf("Error creating a new request for %v: %v", requestURL, err)
+		return nil, err
+	}
+	req.Close = true
+	req.Header.Set("Content-Type", "application/json")
+
+	// Add query parameters
+	q := req.URL.Query()
+	if timestamp == nil {
+		q.Add("market_data", "true")
+		q.Add("localization", "false")
+		q.Add("tickers", "false")
+		q.Add("community_data", "false")
+		q.Add("developer_data", "false")
+	} else {
+		timestampFormatted := timestamp.Format(cg.timeFormat)
+		q.Add("date", timestampFormatted)
+	}
+	req.URL.RawQuery = q.Encode()
+
+	client := &http.Client{
+		Timeout: cg.httpTimeoutSeconds,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New("Invalid response status: " + string(resp.Status))
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return bodyBytes, nil
+}
+
+// GetData gets fiat rates from coingecko.com at the specified date and returns JSON string.
+// If timestamp is nil, it will download the latest market data.
+func (cg *CoingeckoDownloader) GetData(timestamp *time.Time) (string, error) {
```

Comment:
Why do you Unmarshal->Marshal and then Unmarshal (by the caller) again?

Reply:
Unmarshal extracts the specific field (a map of tickers) from API response, and Marshal packs it into a nice JSON string to be stored in DB. I didn't come up with a better way to do it :|

Reply:
Resolved IRL